pkg/controller/machinepool/machinepool_controller.go

```diff
@@ -581,19 +581,25 @@ func (r *ReconcileMachinePool) syncMachineSets(
 					}
 				}
 
-				// Update if the labels on the remote machineset are different than the labels on the generated machineset.
+				// Merge the labels on the remote machineset, if different from the labels on the generated machineset.
 				// If the length of both labels is zero, then they match, even if one is a nil map and the other is an empty map.
-				if rl, l := rMS.Spec.Template.Spec.Labels, ms.Spec.Template.Spec.Labels; (len(rl) != 0 || len(l) != 0) && !reflect.DeepEqual(rl, l) {
+				if rl, l := rMS.Spec.Template.Spec.Labels, ms.Spec.Template.Spec.Labels; len(rl) != len(l) ||
+					(len(rl) == len(l) && !reflect.DeepEqual(rl, l)) {
```

Comment:
I think we will need to test this by checking to see if the generated labels are a subset of the remote labels otherwise we will always report a difference / perform an update even if the generated labels were already present on the remote machineset. Same will apply to taints.

Reply:
So - generated labels and taints first import the same from MachinePool and then should preserve what is already present on the remote MachineSet. Do I understand that correctly?
If so - I added another test that ensures an update is not called when the labels/taints are already present - even if out of order (though reflect.DeepEqual doesn't really care about order so that was a moot point)

Reply:
Yes, we should determine that an update needs to be made if the labels we expect aren't in the existing labels.